### ISSUE TITLE
Support Spontaneous Scopes for UMA 2

### ISSUE THREAD

**Issue by yuriyz** [GluuFederation/oxAuth#1133]
## Problem description

We got [support ticket](https://support.gluu.org/access-management/7344/fine-grained-api-protection-with-uma/#at50863) where customer wish to grant access to path based on user.

```
/user/1
...
/user/n
```

To support this use case for 1000 000 users we need to generate 1000 000 scopes (so user with scope `/user/1` will have access only to `/user/1` and will be rejected if will try to access `/user/2`).

The problem with this approach is that we have to create as much scopes as we have users (thus different paths).

## Idea how to solve it

We can solve it with #839. In this case we don’t need 1 000 000 scopes. We can enable Spontaneous scope for given client. In this case UMA engine on AS side should be smart enough to provide appropriate validation and scopes releasing for such scopes.

**Comment by nynymike** [comment]
What about the idea for Spontaneous Scopes: https://github.com/GluuFederation/oxAuth/issues/839

**Comment by yuriyz** [comment]
@nynymike right, so it spontaneous scopes for UMA 2. I will change title to keep it in sync.

**Comment by yuriyz** [comment]
Main question during implementation is how to map UMA Policy to non-existent. Therefore we will use scope with place holder, like `users/{}` or `pis:{}`.

- RS requests particular scope (e.g. `users/1`) which should match registered scope template.
- We can associate policy with scope with placeholder
- During permission registration we will create real scope and persist it. We need real scope saved to have correct reference in RPT permissions (so we differentiate between `users/1` and `users/2`).

On implementation level we will have boolean indicator which says whether it’s spontaneous scope.

**Comment by altexy** [comment]
@yuriyz can you, please, provide detailed example spontaneous scope workflow with actors below:
- UMA client
- GG as UMA PEP
- oxd
- GS

The puzzle is not completed in my mind.

0. `/users/?/??` path (use our wildcard syntax) is protected with spontaneous scope `/users/{}`
1. UMA client tries to access `/user/2/command`
2. GG detect the protected resource, call oxd.check_access() without RPT
3. Which spontaneous scope oxd should register? The rules of mapping URL to scopes may be very different.

**Comment by yuriyz** [comment]
Lets say we have following paths on RS (GG) :
```
/users/1/command
/users/2/command
...
/users/n/command
```

RS (GG) protects path with scopes. For this use case RS (GG) registers resource (`resource_users`) with path `/users` and scope `/users/{}` (oxd `/uma-rs-protect`). On GS (oxauth) we map policy to `/users/{}`.

Now flow begins
1. RP requests `/users/2/command` on RS (GG)
2. RS (GG) send to oxd `/uma-rs-check-access` with scope `/users/2` (requested scope must match scope template registered during resource registration)
3. since RPT is not authorized, permission is registered with `/users/2` and ticket returned
4. RP requests RPT with ticket (policy is invoked on oxauth and on success) get back RPT authorized with permission for `resource_id: "resource_users"` and scope `/users/2`.
5. RP sends request to RS (GG), GG see `/users/2` scope in RPT, means RP can access path to `/users/2/command`.

For more complex examples handling would be the same:
```
/users/1/command/1/inner
/users/1/command/2/inner
...
/users/1/command/m/inner
...
/users/2/command/1/inner
/users/2/command/2/inner
...
/users/2/command/m/inner
...
/users/n/command/1/inner
/users/n/command/2/inner
...
/users/n/command/m/inner
```

1. resource = `resource_users`, scope = `/users/{}/command/{}`.
2. policy attached to `/users/{}/command/{}` 
3. `/uma-rs-check-access` with scope e.g. `/users/2/command/3`.
4. Policy invoked and authorized RPT with scope `/users/2/command/3` for resource `resource_users`.
5. Access to path `/users/2/command/3/inner` is granted.

**Comment by altexy** [comment]
@yuriyz thanks for the detailed example.

>RS (GG) protects path with scopes. For this use case RS (GG) registers resource (resource_users) with path /users and scope /users/{} (oxd /uma-rs-protect). On GS (oxauth) we map policy to /users/{}.

We don't use prefix match in GG 4.0, we use our wildcard/regexp notation.
`/users/?/??` means:
1. Start with '/users
2. Has exactly one subfolder (?) with any allowed content
3. Has any suffix (??) may contain subfolders

Also we may use regex for a single path element using {} notation, for example:
'/users/{^\d+$}/??` - here we require that userId must be one or more digits.

oxd doesn't need to interpret this notation in any way.

How many dynamic scopes a protected resource may have? Do we need more than one?
IMO it may be a powerful feature.
**GS must be able to match the dynamic scope to the template.**

So we need a convention, one possible convention is to reuse our wildcard/regex syntax.

But should we use prefix like approach for dynamic scope template?
Mike's examples from here https://github.com/GluuFederation/oxAuth/issues/839#issue-337321479 inspire me to use not based on uri path prefix template, but just PCRE regex.

For example:
We have the next dynamic scopes attached to a resource:
1. `^users:\d+$`
2. `^command:(delete|bill)$`
3. `^command:\.+$`

Client tries to access `/users/2/command/view` page.
GG parses the path by some of its own rules and request 2 scopes: `users:1` and command:view`
GS match scope 1 and 3 and invoke attached policies.

Client tries to access `/users/2/command/delete` page.
GG request 2 scopes: `users:1` and command:delete`
GS match scope 1 and 2.

The example shows that the scopes regex order maybe important.

Of course we may implement one dynamic complex scope like from @yuriyz example:
`^\/users\/\d+\/command\/\.+$`

**Comment by yuriyz** [comment]
Right, we can match many dynamic scopes. And agreed, lets use regexp on GS (oxauth) side.

**Comment by nynymike** [comment]
anything can be expressed in regexp, so why not.

**Comment by yuriyz** [comment]
Implemented in 4.2 (docs added).
